src/NitroSharp/Content/AFSFile.cs

```diff
@@ -0,0 +1,170 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.MemoryMappedFiles;
+using System.Linq;
+using System.Text;
+
+namespace NitroSharp.Content
+{
+    internal class AFSFile : IArchiveFile
+    {
+        public static readonly byte[] Magic = {0x41, 0x46, 0x53, 0x00};
+
+        private uint _entriesCount;
+
+        private (uint offset, uint size)[]? _files;
+        private readonly Dictionary<string, uint> _builtinFileNames;
+        private readonly Dictionary<string, uint> _iniFileNames;
+
+        protected readonly MemoryMappedFile _mmFile;
+        protected readonly uint _archiveOffset;
+
+        private readonly Encoding _encoding;
+
+        private bool _disposed = false;
+
+        public AFSFile(MemoryMappedFile mmFile, Encoding encoding)
+        {
+            _builtinFileNames = new Dictionary<string, uint>();
+            _iniFileNames = new Dictionary<string, uint>();
+            _encoding = encoding;
+            _mmFile = mmFile;
+            _archiveOffset = 0;
+        }
+
+        public AFSFile(AFSFile parent, string name, Encoding encoding)
+        {
+            _builtinFileNames = new Dictionary<string, uint>();
+            _iniFileNames = new Dictionary<string, uint>();
+            _encoding = encoding;
+            _mmFile = parent._mmFile;
+            _archiveOffset = parent._archiveOffset + parent.GetFile(name).offset;
+        }
+
+        public static IArchiveFile Create(MemoryMappedFile mmFile, Encoding encoding)
+        {
+            return new AFSFile(mmFile, encoding);
+        }
+
+        ~AFSFile()
+        {
+            Dispose(false);
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        public virtual void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                if (disposing && _archiveOffset == 0)
+                {
+                    _mmFile?.Dispose();
+                }
+                _disposed = true;
+            }
+        }
+
+        public void OpenArchive()
```

Comment:
I've fixed the issue but I had to leave an `if (_files == null)` in `AFSFile.GetFile` overwise the compiler will continue to issue warnings. Maybe we should not use an ArchiveException in this case. Is there any exception that looks like "this is absolutly not normal that we got here" ?

Reply:
It's still theoretically possible for ``AFSFile`` to be in an invalid state since ``_files`` is a) nullable b) mutable. As far as the compiler is concerned, you haven't fixed a thing 🙂 
There's nothing preventing you from initializing it in the constructor like this:
```csharp
private AFSFile(
    MemoryMappedFile mmFile,
    uint offset,
    Encoding encoding,
    (uint offset, uint size)[] files,
    Dictionary<string, uint> builtinFileNames)
{
    _iniFileNames = new Dictionary<string, uint>();
    _mmFile = mmFile;
    _archiveOffset = offset;
    _encoding = encoding;
    _files = files;
    _builtinFileNames = builtinFileNames;
}
```

Notice that this constructor takes ``offset`` as a parameter. There's really no need to have 2 almost identical constructors if you can call this one with ``(parent._mmFile, parent._archiveOffset + parent.GetFile(name).offset, encoding, files, builtinFileNames)``

As for the ``OpenArchive`` method, it's longer necessary for it to exist. Conceptually, it's something that happens as part of *loading* the archive, right? You can have one TryLoad method that takes care of everything:
```csharp
public static AFSArchive? TryLoad(MemoryMappedFile mmFile, Encoding encoding, uint archiveOffset = 0)
{
    // contents of OpenArchive
    if (!magic.SequenceEqual(Magic))
    {
        return null;
    }
    return new AFSArchive(mmFile, archiveOffset, encoding, files, builtinFileNames);
}
```

As well as a Load method that throws an exception:
```csharp
public static AFSFile Load(MemoryMappedFile mmFile, Encoding encoding, uint archiveOffset = 0)
{
    return TryLoad(mmFile, encoding, archiveOffset) is AFSFile file
        ? file
        : throw new ArchiveException("AFS", "Unknown magic");
}
```

Reply:
I see, it's sometimes hard for me to wrap my head around the OOP mindset.

I modified `NPAFile` in the same way to keep a symmetry between the formats.
In `NPAFile` we can throw an exception for multiple reasons, I've setup `TryLoad` to only return `null` when the magic is invalid because that should be the only *"normal"* case, the others represent malformed or unsupported files.

Reply:
I basically asked you to make illegal states unrepresentable, that's not exactly an OOP thing 🙂